components/script/dom/element.rs

```diff
@@ -1980,6 +1994,27 @@ impl Element {
             document.perform_focus_fixup_rule(self);
         }
     }
+
+    pub fn get_element_internals(&self) -> Option<DomRoot<ElementInternals>> {
+        self.rare_data().as_ref().and_then(|r| {
+            r.element_internals
+                .as_ref()
+                .map(|a_i| DomRoot::from_ref(&**a_i))
+        })
+    }
+
+    pub fn ensure_element_internals(&self) -> DomRoot<ElementInternals> {
+        let mut rare_data = self.ensure_rare_data();
+        if let Some(internals) = &rare_data.element_internals {
+            return DomRoot::from_ref(&*internals);
+        }
+        let elem = self
+            .downcast::<HTMLElement>()
+            .expect("ensure_element_internals should only be called for an HTMLElement");
+        let internals = ElementInternals::new(elem);
+        rare_data.element_internals = Some(Dom::from_ref(&*internals));
+        internals
```

Comment:
I think that `get_or_insert_with(...)` works well here:

```suggestion
        let mut rare_data = self.ensure_rare_data();
        DomRoot::from_ref(rare_data.element_internals.get_or_insert_with(|| {
            let elem = self
                .downcast::<HTMLElement>()
                .expect("ensure_element_internals should only be called for an HTMLElement");
            Dom::from_ref(&*ElementInternals::new(elem))
        }))
```